kotlin/workflow-ui-android/src/main/java/com/squareup/workflow/ui/WorkflowRunnerViewModel.kt

```diff
@@ -20,70 +20,64 @@ import androidx.lifecycle.ViewModel
-import kotlinx.coroutines.flow.emitAll
-import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.launchIn
 import kotlinx.coroutines.flow.map
-import kotlinx.coroutines.launch
-import kotlinx.coroutines.rx2.asFlowable
+import kotlinx.coroutines.flow.onEach
 import kotlinx.coroutines.rx2.asObservable
+import kotlinx.coroutines.rx2.rxSingle
 import org.jetbrains.annotations.TestOnly
 import java.util.concurrent.CancellationException
 import kotlin.reflect.jvm.jvmName
 
 @UseExperimental(ExperimentalCoroutinesApi::class)
 @ExperimentalWorkflowUi
 internal class WorkflowRunnerViewModel<OutputT : Any>(
-  override val viewRegistry: ViewRegistry,
-  private val renderingsFlow: Flow<RenderingAndSnapshot<Any>>,
-  outputsFlow: Flow<OutputT>,
-  private val scope: CoroutineScope
+  private val scope: CoroutineScope,
+  renderingsFlow: Flow<RenderingAndSnapshot<Any>>,
+  output: Flow<OutputT>,
+  override val viewRegistry: ViewRegistry
 ) : ViewModel(), WorkflowRunner<OutputT> {
 
-  /**
-   * @param inputs Function that returns a channel that delivers input values for the root
-   * workflow. The first value emitted is passed to `initialState` to determine the root
-   * workflow's initial state, and subsequent emissions are passed as input updates to the root
-   * workflow. The channel returned by this function will be cancelled by the host when it's
-   * finished.
-   */
   @UseExperimental(ExperimentalCoroutinesApi::class)
   internal class Factory<InputT, OutputT : Any>(
-    private val workflow: Workflow<InputT, OutputT, Any>,
-    private val viewRegistry: ViewRegistry,
-    private val inputs: Flow<InputT>,
     savedInstanceState: Bundle?,
-    private val dispatcher: CoroutineDispatcher = Dispatchers.Main.immediate
+    private val configure: () -> WorkflowRunner.Config<InputT, OutputT>
   ) : ViewModelProvider.Factory {
     private val snapshot = savedInstanceState
         ?.getParcelable<PickledWorkflow>(BUNDLE_KEY)
         ?.snapshot
 
-    override fun <T : ViewModel> create(modelClass: Class<T>): T = launchWorkflowIn(
-        CoroutineScope(dispatcher),
-        workflow,
-        inputs,
-        snapshot
-    ) { renderings, outputs ->
-      @Suppress("UNCHECKED_CAST")
-      WorkflowRunnerViewModel(viewRegistry, renderings, outputs, this) as T
+    override fun <T : ViewModel> create(modelClass: Class<T>): T {
+      return with(configure()) {
+        launchWorkflowIn(
+            CoroutineScope(dispatcher), workflow, inputs, snapshot
+        ) { renderings, output ->
+          @Suppress("UNCHECKED_CAST")
+          WorkflowRunnerViewModel(this, renderings, output, viewRegistry) as T
+        }
+      }
     }
   }
 
-  private val snapshotJob = scope.launch {
+  override val result: Single<out OutputT> = scope
+      .rxSingle { output.first() }
+      .doAfterTerminate {
+        scope.cancel(CancellationException("WorkflowRunnerViewModel delivered result"))
+      }
```

Comment:
I think we should add a `.cache()` after this, so multiple subscribers to the `Single` don't end up trying to cancel the scope multiple times.